### ISSUE TITLE
How to specify a schema

### ISSUE THREAD

**Issue by weltschmerz1** [replikativ/datahike#4]
I would like to use a schema the same way datascript offers [here](https://github.com/tonsky/datascript#usage-examples):

```clojure
(let [schema {:aka {:db/cardinality :db.cardinality/many}}]
  (d/create-conn schema))
```

How would I go about that?

I am basically porting my application from datascript to datahike so that my app has persistent data, but I have certain schematic requirements such as cardinality-many, refs, etc.

**Comment by whilo** [comment]
I have pushed a version to master which should allow you to pass your schema to datahike in a similar way to `create-conn`. This is not ideal, as it does not follow Datomic's schema as data concept and it is not possible yet to change it later. Is this sufficient for you for now?

**Comment by weltschmerz1** [comment]
That looks perfect. Please let me know when you publish it and I'll try it out. Thank you so much!

**Comment by whilo** [comment]
You can use the `0.1.1-SNAPSHOT` version for now.

**Comment by weltschmerz1** [comment]
I tried it out, but I don't think it's working yet. Either I am not loading the schema correctly, or the schema is not taking effect.

Perhaps one indicator that the schema is not being loaded correctly is when I print the value of my `conn` atom, it prints nil for its `:schema`, even though I did use the new `create-database-with-schema`.

On the off chance that this is expected or incidental, let me elaborate on the problem I am seeing:

For fields that are refs with cardinality many, transacting multiple refs at once works fine. However, running a transaction with the intention of adding a second ref to an existing field has no effect; the ref does not get added.

For example, let's say our schema looks like this:

```edn
{:field {:db/valueType   :db.type/ref
         :db/cardinality :db.cardinality/many}}
```

Then writing a transaction that looks like this

```clojure
[{:field [1 2]}]
```

works fine; `:field` indeed has refs to entities 1 and 2.

This transaction, however

```clojure
[[:db/add x :field 1]
 [:db/add x :field 2]]
```

(x being some entity id that can be resolved) has the effect of only adding 2. It's almost as if 2 ended up overriding 1.

I'll summarize this by saying again that most likely it's a symptom of the schema not actually taking effect, which probably spares you from going down the rabbit hole of trying to reproduce this particular example. I can reproduce the same behavior when running the previous version, `0.1.0`.

**Comment by whilo** [comment]
I am sorry, I should have written a test in the first place. It was not passed through serialization, it should be fixed now and is covered with a test from datascript. The new SNAPSHOT version is on clojars.

**Comment by weltschmerz1** [comment]
Another datapoint: Upon running a query after connecting to an existing database, I now get this exception:

```clojure
#error {
 :cause nil
 :via
 [{:type java.lang.NullPointerException
   :message nil
   :at [datahike.db.DB _attrs_by db.cljc 503]}]
 :trace
 [[datahike.db.DB _attrs_by db.cljc 503]
  [datahike.db$is_attr_QMARK_ invokeStatic db.cljc 927]
  [datahike.db$is_attr_QMARK_ invoke db.cljc 925]
  [datahike.db$ref_QMARK_ invokeStatic db.cljc 935]
  [datahike.db$ref_QMARK_ invoke db.cljc 933]
  [datahike.pull_api$pull_attr_datoms invokeStatic pull_api.cljc 116]
  [datahike.pull_api$pull_attr_datoms invoke pull_api.cljc 109]
  [datahike.pull_api$pull_expand_frame invokeStatic pull_api.cljc 196]
  [datahike.pull_api$pull_expand_frame invoke pull_api.cljc 189]
  [datahike.pull_api$pull_wildcard_expand invokeStatic pull_api.cljc 220]
  [datahike.pull_api$pull_wildcard_expand invoke pull_api.cljc 210]
  [datahike.pull_api$pull_wildcard invokeStatic pull_api.cljc 226]
  [datahike.pull_api$pull_wildcard invoke pull_api.cljc 222]
  [datahike.pull_api$pull_pattern_frame invokeStatic pull_api.cljc 232]
  [datahike.pull_api$pull_pattern_frame invoke pull_api.cljc 228]
  [datahike.pull_api$pull_pattern invokeStatic pull_api.cljc 254]
  [datahike.pull_api$pull_pattern invoke pull_api.cljc 249]
  [datahike.pull_api$pull_spec invokeStatic pull_api.cljc 269]
  [datahike.pull_api$pull_spec invoke pull_api.cljc 266]
  [datahike.query$pull$iter__24110__24114$fn__24115$fn__24121 invoke query.cljc 740]
  [clojure.core$map$fn__4789 invoke core.clj 2651]
  [clojure.lang.LazySeq sval LazySeq.java 40]
  [clojure.lang.LazySeq seq LazySeq.java 49]
  [clojure.lang.RT seq RT.java 521]
  [clojure.core$seq__4357 invokeStatic core.clj 137]
  [clojure.core.protocols$seq_reduce invokeStatic protocols.clj 24]
  [clojure.core.protocols$fn__6738 invokeStatic protocols.clj 75]
  [clojure.core.protocols$fn__6738 invoke protocols.clj 75]
  [clojure.core.protocols$fn__6684$G__6679__6697 invoke protocols.clj 13]
  [clojure.core$reduce invokeStatic core.clj 6545]
  [clojure.core$into invokeStatic core.clj 6610]
  [clojure.core$mapv invokeStatic core.clj 6618]
  [clojure.core$mapv invoke core.clj 6618]
  [datahike.query$pull$iter__24110__24114$fn__24115 invoke query.cljc 737]
  [clojure.lang.LazySeq sval LazySeq.java 40]
  [clojure.lang.LazySeq seq LazySeq.java 49]
  [clojure.lang.RT seq RT.java 521]
  [clojure.core$seq__4357 invokeStatic core.clj 137]
  [clojure.core$print_sequential invokeStatic core_print.clj 46]
  [clojure.core$fn__6072 invokeStatic core_print.clj 153]
  [clojure.core$fn__6072 invoke core_print.clj 153]
  [clojure.lang.MultiFn invoke MultiFn.java 233]
  [clojure.core$pr_on invokeStatic core.clj 3572]
  [clojure.core$pr invokeStatic core.clj 3575]
  [clojure.core$pr invoke core.clj 3575]
  [clojure.lang.AFn applyToHelper AFn.java 154]
  [clojure.lang.RestFn applyTo RestFn.java 132]
  [clojure.core$apply invokeStatic core.clj 646]
  [clojure.core$pr invokeStatic core.clj 3588]
  [clojure.core$pr doInvoke core.clj 3575]
  [clojure.lang.RestFn applyTo RestFn.java 139]
  [clojure.core$apply invokeStatic core.clj 646]
  [clojure.core$prn invokeStatic core.clj 3612]
  [clojure.core$prn doInvoke core.clj 3612]
  [clojure.lang.RestFn applyTo RestFn.java 137]
  [clojure.core$apply invokeStatic core.clj 646]
  [clojure.core$println invokeStatic core.clj 3631]
  [clojure.core$println doInvoke core.clj 3631]
  [clojure.lang.RestFn invoke RestFn.java 421]
  [my-project.core$socket_handler$fn__27643$fn__27649 invoke core.clj 133]
  [my-project.core$socket_handler$fn__27643 invoke core.clj 131]
  [manifold.stream.Callback put stream.clj 454]
  [manifold.stream.graph$async_send invokeStatic graph.clj 78]
  [manifold.stream.graph$async_send invoke graph.clj 74]
  [manifold.stream.graph$async_connect$this__2615 invoke graph.clj 209]
  [manifold.stream.graph$async_connect$this__2615 invoke graph.clj 188]
  [clojure.core$trampoline invokeStatic core.clj 6064]
  [clojure.core$trampoline invoke core.clj 6053]
  [manifold.stream.graph$async_connect invokeStatic graph.clj 183]
  [manifold.stream.graph$async_connect invoke graph.clj 125]
  [manifold.stream.graph$connect invokeStatic graph.clj 328]
  [manifold.stream.graph$connect invoke graph.clj 300]
  [manifold.stream$connect invokeStatic stream.clj 330]
  [manifold.stream$connect invoke stream.clj 300]
  [manifold.stream$eval10813$consume__10814 invoke stream.clj 480]
  [my-project.core$socket_handler invokeStatic core.clj 103]
  [my-project.core$socket_handler invoke core.clj 95]
  [my-project.core$handler invokeStatic core.clj 199]
  [my-project.core$handler invoke core.clj 195]
  [clojure.lang.Var invoke Var.java 379]
  [ring.middleware.reload$wrap_reload$fn__15339 invoke reload.clj 39]
  [ring.middleware.cors$handle_cors invokeStatic cors.clj 161]
  [ring.middleware.cors$handle_cors invoke cors.clj 153]
  [ring.middleware.cors$wrap_cors$fn__14810 invoke cors.clj 178]
  [aleph.http.server$handle_request$fn__13200$f__640__auto____13201 invoke server.clj 157]
  [clojure.lang.AFn run AFn.java 22]
  [io.aleph.dirigiste.Executor$Worker$1 run Executor.java 62]
  [manifold.executor$thread_factory$reify__522$f__523 invoke executor.clj 44]
  [clojure.lang.AFn run AFn.java 22]
  [java.lang.Thread run Thread.java 745]]}
```

**Comment by whilo** [comment]
Ok, can you open a separate issue? This does not look related. 

Schema updates are not supported by datascript right away. You can update the schema yourself, as the database is just a hash-map under the `:db` key in the underlying konserve store, but I think this can have unintended consequences. If I understand correctly then datascript recommends reinitializing the database with a new schema. That should definitely be doable as well (although this is not good for datahike ofc. as it can hold potentially a lot more data than datascript). I have just fixed transact as well, I definitely need to take some hammock time for this, I am just busy with other stuff. Ideally we would like to have the schema behaviour of datomic.

**Comment by weltschmerz1** [comment]
Sure I can open a separate issue. It seems to me to be related though, because I'm not updating the schema. I'm simply trying to connect to an existing database that was given a schema when it was created.

Otherwise, using a schema means losing support for persistence, or am I missing something?

In the meantime, if there is a workaround, I'm happy to try it out.

**Comment by weltschmerz1** [comment]
(To be clear, I created a new database with a schema from scratch, and then tried to reuse it.)

**Comment by whilo** [comment]
Ok, hopefully I have time to fix this seriously till the weekend.

**Comment by pm4r** [comment]
I have same problem with datahike:file db. When I create database with schema, release connection and then connect again, queries return no data, just like db would be empty. Actually when looking into connection atom, there is :schema but :datoms are empty.

Create->connect->release->connect works fine if I don't use schema. :datoms are there and :schema is nil in that case.

**Comment by whilo** [comment]
I have difficulties to reproduce, can you retry with the current SNAPSHOT release? Just to be sure I have not pushed an out of date version.

**Comment by pm4r** [comment]
I checked with current SNAPSHOT and problem still exists for ma use case. So I narrowed it down to create simplest case when it occurs. It turns out it's not related with schema but datatypes, specifically bigdec which I was trying to store in db. It worked without complains, and data is querable but after reconnect all data is unavailable.

It works same whether I use schema or not. 

```
(def uri (str "datahike:file://" (.getAbsolutePath (java.io.File. "testdb1"))))
(def data [{:my/prop 42.5M}]) ;42,5, "abc", #inst ... - works well, but bigdec doesn't

(d/create-database uri)
(def conn (d/connect uri))
(d/transact conn data)
(d/q '[:find [(pull ?e [*]) ...]
       :where [?e :my/prop]]
  @conn) ; all fine, our data is there
(type (:my/prop (first *1))) ; -> java.math.BigDecimal

(d/release conn)

(def conn (d/connect uri))
(d/q '[:find [(pull ?e [*]) ...]
       :where [?e :my/prop]]
  @conn) ; empty
```
Change 42.5M to 42.5 and it is available after reconnect.

Looks like storing single bigdec corrupts database somehow, without throwing exception.

**Comment by pm4r** [comment]
Testing further it turns out that storing bigdec while using same connection causes nothing to be stored after it. After connecting again all data after inserting bigdec are missing but I can again insert new values, and :db/id of next entity has value like it never did any transactions since first time inserting bigdec.

So it looks like inserting bigdec causes something like detaching physical storage from memory data, which can be fixed after reconnection, but this causes in-memory data to be lost.

**Comment by whilo** [comment]
@pm4r
Thanks a lot for reporting this! I have not had time to to use datahike enough to hit this issue yet. I think this is a serialization problem that I have not anticipated and somehow slips through, in general datahike only supports core edn types, not custom ones (like BigDecimals). Can you open a separate issue please?

**Comment by whilo** [comment]
@weltschmerz1 I have difficulties to reproduce your schema is nil issue with the current datahike version, but I will track it down. Can you verify it is still happening with the current SNAPSHOT release?

**Comment by pm4r** [comment]
Ok, I created #5 for this.

**Comment by whilo** [comment]
Thank you!

**Comment by pm4r** [comment]
I use database with schema, with various types of values, refs, also with `cardinality/many` and don't see any issues with reconnecting - all data is available. I would suspect @weltschmerz1 encountered that silent problem as me with previous version (solved in #5).

**Comment by whilo** [comment]
Ok, I close it for now. Feel free to reopen it, if the issue persists.
